### ISSUE TITLE
Serve tiles over WebP or AVIF

### ISSUE THREAD

**Issue by grischard** [openstreetmap/operations#902]
[95% of browsers](https://caniuse.com/webp) support WebP, and [84% support avif](https://caniuse.com/avif). By serving our raster tiles over WebP and/or AVIF, we could reduce our bandwidth usage and speed up our maps without a noticeable loss in quality.

Since compatibility is so wide-spread, we should directly render WebP or AVIF tiles instead of PNG, and transcode the tiles to PNG on demand. I'm not sure exactly where in the stack this would happen?

As we would serve image formats depending on the client's support, we need to ensure our caching is done on '<a href="https://blog.cloudflare.com/vary-for-images-serve-the-correct-images-to-the-correct-browsers/">Vary Accepts</a>' headers, both for the CDN and the clients.

The old tile URL would be backwards compatible, but announcing this would be a good opportunity to encourage people to use https://tile.openstreetmap.org/z/x/y (see also #737)

Testing on:

* https://tile.openstreetmap.org/0/0/0.png : PNG 6.8K, WebP 2.9K, AVIF 2.0K
* https://tile.openstreetmap.org/14/8470/5583.png : PNG 36K, WebP 18K, AVIF 13K
* https://tile.openstreetmap.org/14/12307/7078.png : PNG 44K, WebP 19K, AVIF 15K

**Comment by Firefishy** [comment]
Mapnik supports generating WebP natively.

**Comment by grischard** [comment]
https://github.com/mapnik/mapnik/wiki/Image-IO says webp is experimental.

AVIF support is "[future](https://github.com/mapnik/mapnik/issues/4315)".

Can fastly auto-transcode webp to png?

**Comment by pnorman** [comment]
I would be reluctant to build something that relies on Fastly's closed-source components. Additionally, I would want to serve webp and transcode it to PNG, which would require mod_tile to support webp metatiles.

**Comment by pnorman** [comment]
I wouldn't mind moving to all webp - metatiles in webp, sliced into either png or webp tiles.

Looking at https://jpegxl.info/comparison.png, webp is considered better for lossless non-photographic, while avif is better for mixed photo/nonphoto, given what Netflix designed it for. Of course, that particular chart recommends JPEG XL since it's from the JPEG XL community site and has an agenda. JPEG XL isn't an option with virtually no support among browsers.

But we'd still need to produce PNG tiles for the significant number of PNG-only users that aren't browsers.

I don't see this being worth the effort, given the state of mod_tile and other priorities for developing the standard layer

**Comment by grischard** [comment]
If the mod_tile route is a dead end, let's see about using Fastly's Image Optimizer? We prefer open source, of course, but only if it's possible. Apparently, Image Optimizer is an extra service - can we ask them if they'd be willing to enable it on our account?

Even if we go all-out on vector tiles, I expect that we will still be serving raster tiles for a few years.

**Comment by zax71** [comment]
This would be a really useful feature! According to a Lighthouse report, for my website, this could just about half the size of loaded images.

<img width="808" height="398" alt="Image" src="https://github.com/user-attachments/assets/a0110a3e-a710-4b84-a8d3-370f0dff9c71" />

**Comment by hummeltech** [comment]
@zax71, yes `webp` has been supported for quite some time, here's an example of how to enable it:
https://github.com/openstreetmap/mod_tile/blob/f0811e8f298d77e94029425c182425174fcba19f/README.rst?plain=1#L125

**Comment by zax71** [comment]
How would I use that with a `tile.openstreetmap.org` link? Replacing `.png` with `.webp` causes an error. I am currently using `https://tile.openstreetmap.org/{z}/{x}/{y}.png` with Sveaflet

**Comment by tomhughes** [comment]
You can't because it's not enabled. I'm not sure if we even have that support in the code we're running.

**Comment by zax71** [comment]
Ah, so it's not really possible then. Or are there 3rd party hosts for OSM?

**Comment by tomhughes** [comment]
Lots of people create maps using OSM data using all sorts of tools and output formats.

**Comment by zax71** [comment]
I'm just using a standard OSM with Leaflet

**Comment by majudev** [comment]
My testing shows that using lossless WebP or AVIF causes larger outputs, than using png with no alpha and compression at level 9.

**Comment by zax71** [comment]
That’s interesting, any theories as to why that is? Webp and the other modern image formats are *designed* to have really high compression ratios. Especially compared to a lossless format like png

Could it maybe be that the OSM tiles are not “photographs” and hence the compression algorithms are not optimised for this use case? Although I don’t know a thing about this and that’s just a theory.

**Comment by majudev** [comment]
Maybe I was not precise enough. In my testing, baseline was lossless png, no alpha, zlib compression level 9. I selected around 1000 random tiles from Poland on zoom level 12. Results are:
- lossless WebP / lossless AVIF: size **increase** of 1-5%
- avif quality 90, speed 6: size **decrease** of around 30%
- avif quality 80, speed 4: pretty much inline what original author has reported (around 50-60% size **decrease**)

So yes, AVIF and WebP provide an advantage when used in lossy mode, but not in lossless mode and that's what I meant earlier. I did not spot much of a difference when using quality 90 or even 80, but I am not sure if compressing already blurry 256x256 pixel tiles is the best idea. If this change is ever going to be implemented, it shall come with an increase of default pixel size to 512x512 pixels.

As a sidenote, I don't know what png settings OSM uses right now. Just removing alpha channel reduced size substantially.

**Comment by zax71** [comment]
Ah, cool! That looks like really useful data.

**Comment by quantenschaum** [comment]
I did the following test: Using [convert.py](https://github.com/quantenschaum/mapping/blob/master/scripts/convert.py) (which uses Pillow to save the images) I reencoded the [`qmap-de.mbtiles`](https://freenauticalchart.net/download/qmap-de.mbtiles) to png and webp.

```shell
scripts/convert.py charts/qmap-de.mbtiles x.png.mbtiles -f -Z12 -Fpng -CO
scripts/convert.py charts/qmap-de.mbtiles x.webp.mbtiles -f -Z12 -Fwebp -CO
```

(`-Z12`=up to zoom level 12 (for speed), `-C`=force reencode, `-O`=remove transparency)

Tiles where encoded with [Pillow options](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html#png-saving) `lossless=True` and `optimize=True`. 

The results:

```shell
-rw-r--r-- 1 al al 39M Jan 11 20:25 x.png.mbtiles
-rw-r--r-- 1 al al 24M Jan 11 20:29 x.webp.mbtiles
```

So, I disagree on png being smaller than webp.

Keeping transparency (only tiles with actual transparent pixels have alpha channel) I get png=40MB vs webp=24MB.

The full file is 812MB webp vs 1700MB png.

Because map tiles are not photographs with distinct colors and areas with same color, lossless webp is expected to be smaller than png.

@majudev How exactly did you perform your test? I would like to reproduce them.

**Comment by quantenschaum** [comment]
With [Pillow AVIF](https://pypi.org/project/pillow-avif-plugin/) and default settings I get it down to 14MB.

**Comment by majudev** [comment]
Interesting. I updated my scripts to be a bit more verbose and put them in [this repo](https://github.com/majudev/osm-compression-benchmark). It includes source tiles which serve as a baseline (from a personal project).

```
====================================================================================
SUMMARY (zoom = 6)
====================================================================================
Format                              Avg Size        Total Size      vs Original     
------------------------------------------------------------------------------------
Original PNG                        297.33 KB       18.58 MB        baseline        
Re-encoded PNG (no alpha, L9)       307.05 KB       19.19 MB        +3.27%          
AVIF Q=95 S=6 (Fast)                96.24 KB        6.01 MB         -67.63%         
AVIF Q=80 S=4 (Balanced)            28.00 KB        1.75 MB         -90.58%         
AVIF Q=100 S=4 (Lossless)           364.68 KB       22.79 MB        +22.65%         
====================================================================================
```

**Comment by gravitystorm** [comment]
I recommend testing **only** the map tiles that hosted by the OSMF and served from `tile.openstreetmap.org`.

Every map style is different, they all have different number of colours and different information density, and the differences can be substantial when it comes to image file sizes and compressibility. Whether or not a completely different style compresses better with webp doesn't necessarily prove anything about this specific tile layer.

**Comment by imagico** [comment]
> Every map style is different, they all have different number of colours and different information density, and the differences can be substantial when it comes to image file sizes and compressibility.

Indeed - furthermore, it depends also on what you want to optimize for, whether that is tile storage volume or user bandwidth use.  For OSM-Carto, for example, compression efficiency of different formats depends a lot on the content of the tile.  If you have a near empty tile or one with lots of visible data.  A quick example:

tile | png size (as served) | webp size (from `magick -quality 100`)
-- | -- | --
[11947](https://tile.openstreetmap.org/15/16820/11947.png) | 31002 | 28136
[53668](https://tile.openstreetmap.org/17/71963/53668.png) | 103 | 44
[53669](https://tile.openstreetmap.org/17/71964/53669.png) | 758 | 346
[53670](https://tile.openstreetmap.org/17/71963/53670.png) | 103 | 44

In this case: empty or near empty OSM-Carto are less than half the size as webp, for content rich tiles the difference is much smaller.  Needless to say that lossy compression on empty or near empty tiles is counterproductive, even from a purely technical standpoint.

**Comment by quantenschaum** [comment]
Yes, it depends on the actual tile content. Standard OSM tiles from https://www.openstreetmap.org/ are **indexed** PNGs So, technically they are lossy using 256 colors only, this is why they get so small. But for this purpose it's fine, it looks good. For other tiles, especially with transparency or antialiasing, 256 color are not enough, but full RGB are way bigger.  Also the palette has to be tuned/fixed for consistent colors across tiles.

I reran a test on actual Standard OSM tiles from around Berlin, zooms 8,10,12,14, 20x20 tiles each.

```
webp       28.41 MB    -10%
png-P      30.76 MB     -2%
orig       31.40 MB     +0%
avif       47.83 MB    +52%
png-RGB    76.73 MB   +144%
```

- indexed PNG with tuned palette is pretty good and very compatible (`png-P`=reencoded with pillow with `mode=P `and `optimize=True`)
- lossless WebP outperforms PNG even for indexed mode
- lossless AVIF is really bad
- RGB PNG is even worse

**Comment by majudev** [comment]
Thank you @quantenschaum for your tests. They are certainly more trustworthy than mine, being done on actual OSM-Carto style. I think it's clear that lossless AVIF is a dead-end.

Expanding on what @imagico pointed - if the goal is to reduce **storage** size, then WebP might be worth trying. But current PNG's seem to be optimized quite well. There is another aspect: time to compress. PNG is instantenous, while WebP might take some time. In case of lossless AVIF, time difference was noticeable, although I did take any precise measurements. Benchmark on actual implementation of mapnik's WebP output would be incredibly useful.

On the other hand, if the goal is to reduce bandwidth on client side, then I don't think it's worth it. Both @quantenschaum and @imagico agree on size reduction of around ~10% on non-empty tiles, and I don't think it will not be noticeable anyway. Empty tiles are very small already, so for bandwidth they don't not really matter.

> Needless to say that lossy compression on empty or near empty tiles is counterproductive, even from a purely technical standpoint.

I agree, with one caveat. Lossy AVIF is faster than lossless AVIF. Noticeably faster. This means that simply setting lossy compression for all tiles, including empty ones, would be more efficient overall. And lossy AVIF can provide ~60% size reduction for content rich tiles - that would be noticeable both if we want to reduce user bandwidth or storage. Most important question regarding lossy AVIF is how to set quality (Q) so that degradation is barely noticeable while providing substantial size reduction. I was not able to spot any degradation in with the tiles compressed using my scripts, but I suppose only objective way to measure it is to partner with some QoE researchers to conduct an experiment.

Overall, I think best option is to either leave it as-is because PNGs already perform okay-ish, or check the feasibility of lossy AVIF/WebP for significant storage benefits. But that requires time and likely money.

**Comment by pnorman** [comment]
Something that hasn't been touched on on is how we actually store tiles and serve them from disk. 8x8 metatiles are rendered in 8 bit palleted PNG and then sliced up into PNGs.

There's a whole bunch of things that could be done. e.g. we could write lossless or lossy metatiles and slice to png8.

Something important is that that you look at files generated by mapnik. When a metatile is written as png8 it reduces the color depth to 8 bit. Is there a way to do the same for webp so it can use the color indexing transform? If not, suddenly the webp metatile contains a lot more information and is unlikely to be smaller. That means fewer tiles in the store which drops the hit rate, resulting in much more work.

I have no idea how much CPU time is spent encoding png8 tiles right now. I suspect it's small compared to the time spent in postgres, mapnik style rules and agg but have no data.

Even if webp metatiles reduce size on disk, I think it would need to be **significant** to be worth it.

**Comment by quantenschaum** [comment]
Thanks to @majudev and @pnorman comments I ran some test with [this script](https://github.com/quantenschaum/mapping/blob/master/scripts/tile-benchmark.py) including lossy formats just for comparison.

OSM tiles around Berlin (200 tiles, zooms 10, 12)
```
webp-80    2.32 MB     -54%    1.1s
jpg-80     2.74 MB     -46%    0.1s
avif-80    3.21 MB     -36%    6.5s
webp-90    3.56 MB     -29%    1.3s
jpg-90     4.07 MB     -19%    0.1s
webp-LL    4.55 MB     -10%    2.4s
avif-90    4.58 MB      -9%    7.9s
png-P7     4.58 MB      -9%    2.9s
png-P8     4.95 MB      -2%    2.9s
gif        5.77 MB     +15%    1.5s
avif-LL    8.26 MB     +64%    9.9s
jpg-100    10.41 MB   +107%    0.3s
png-RGB    12.47 MB   +148%    6.3s
```

So, 8bit PNG is actually very good and much faster to encode than WebP.

I also reencoded all tiles from freenauticalchart.net and found that 8bit PNG is actually better than lossless WebP for the entire dataset, smaller and faster.

But indexed colors can lead to artifacts like this (background should be white, but there are shades of blue)

<img width="844" height="431" alt="Image" src="https://github.com/user-attachments/assets/a637b3fd-2b62-4e31-be4c-836e607da612" />

**Comment by pnorman** [comment]
> So, 8bit PNG is actually very good and much faster to encode than WebP.

Given this I'm going to close the issue. I am curious if metatile results differ from tile results, but not curious enough to leave it open.

If someone wants to do further research they should

- look at 8x8 metatiles rendered by mapnik
- look at tiles written directly to webp without ever going through png8, as they may be larger since they contain more color information
- look at times to slice metatiles into tiles of both the same format and different formats
- look at tile sizes

A WebP with 8-bit color might beat png8 in some ways. I can't see a WebP with 24-bit color beating png8.

The results rule out avif.

> But indexed colors can lead to artifacts like this (background should be white, but there are shades of blue)

I have seen one severe example of this when there were only a few pixels of one color in a metatile and lots of other colors in other parts of the metatile. The few pixels didn't match the color adjacent to it. I don't know if the palletization happens before or after the bounds are trimmed off, but it would influence how this happens.

Given how perceptive our users are about tile quality, we would have reports if it was a wide-spread severe issue.
